### PR TITLE
add forEach to Pool

### PR DESCRIPTION
**Pull request by kevmoo**
Fixes https://github.com/dart-lang/pool/issues/22

### CODE REVIEW

**Comment by natebosch** [comment]
Would the API be nicer as extension methods on Iterable/Stream? And if so, is it with waiting for that option?

**Comment by kevmoo** [comment]
> Would the API be nicer as extension methods on Iterable/Stream? And if so, is it with waiting for that option?

We'd need a general notion of "pooling", though – so I think it makes sense here for now.

Can always upgrade it to an extension method later. This is not a `dart:` API

File: lib/pool.dart [hunk 1]
```diff
@@ -125,6 +125,97 @@ class Pool {
     }
   }
 
+  /// Returns a [Stream] containing the result of [action] applied to each
```
Comment (natebosch) [inline lib/pool.dart]: Please document ordering behavior. Do elements emit on the stream in the order that their `action` completes, or in order that the input appeared in `sourceItems`?

Comment (kevmoo) [reply lib/pool.dart]: done

File: test/pool_test.dart [hunk 2]
```diff
@@ -697,6 +697,11 @@ void main() {
       });
     });
   });
+
+  test("throw error when pool limit <= 0", () {
```
Comment (natebosch) [inline test/pool_test.dart]: Single quotes for consistency with surrounding code

Comment (kevmoo) [reply test/pool_test.dart]: This is merged from a user commit. We should have a follow-up to just move to single quotes.

For now, not specific to this CL

File: lib/pool.dart [hunk 3]
```diff
@@ -130,6 +130,97 @@ class Pool {
...
+
+    Future<void> run() async {
+      while (iterator.moveNext()) {
+        // caching `current` is necessary because there are async breaks
+        // in this code and `iterator` is shared across many workers
+        final current = iterator.current;
+
+        _resetTimer();
+
+        await resumeCompleter?.future;
```
Comment (natebosch) [inline lib/pool.dart]: Probably want to update the doc comment with this caveat - if you pause the stream it'll still hold on to any resources it was using when it was paused, so you can't use the pool for other work while it's paused.

Comment (kevmoo) [reply lib/pool.dart]: fixed

**Comment by kevmoo** [comment]
Note: there is likely a more elegant way to handle this where resources are requested as needed, but to do that we'd need to add knowledge of `forEach` to the core plumbing of `Pool`.

For now, I figured this was a good start – we can evolve the implementation if desired.

**Comment by natebosch** [comment]
@jakemac53 - we had discussed offline that we didn't like how this implementation could starve other users of the `Pool` while the work on the iterable is done. After discussing with @kevmoo this is because it enables us to have another nice property, which is that you can `close()` the pool after listening on the Stream and not get an error on the next item that gets processed.

**Comment by kevmoo** [comment]
@natebosch – so...LGTY?

File: test/pool_test.dart [hunk 4]
```diff
@@ -437,6 +437,288 @@ void main() {
...
+          Iterable<int>.generate(40, (i) {
+            expect(generatedCount, lessThanOrEqualTo(dataCount + 2 * poolSize),
+                reason: 'The iterator should not be called '
+                    'much faster than the data is consumed.');
+            generatedCount++;
+            return i;
+          }),
+          delayedToString);
+
+      // ignore: cancel_subscriptions
```
Comment (natebosch) [inline test/pool_test.dart]: maybe drop the lint?

Comment (kevmoo) [reply test/pool_test.dart]: If a lint is 98% right, I'd rather keep it and comment on the exceptions

**Pull request merged**